Domain filter pattern support

jrchudy opened the pull request:
This PR addresses issue #344. This PR adds a function to the reference object, `filteredRef`, that returns a reference based on a defined `domain-filter-pattern` for the `foreign-key` annotation or the original reference if there is no such annotation defined. This PR needs to be merged before [this PR](https://github.com/informatics-isi-edu/chaise/pull/1020) in `chaise`.

RFSH commented on js/reference.js:
```diff
@@ -1516,6 +1516,30 @@ var ERMrest = (function(module) {
+         * @param {Object} data - tuple data with potential constraints
+         * @returns {ERMrest.Reference} the constrained reference
+         */
+        filteredRef: function(column, data) {
```
`filteredRef` should be a property of [`ReferenceColumn`](https://github.com/informatics-isi-edu/ermrestjs/blob/master/js/reference.js#L2579) not `Reference`.

jrchudy replied:
You are correct. I will move this.

RFSH commented on js/reference.js:
```diff
@@ -1516,6 +1516,30 @@ var ERMrest = (function(module) {
+            if (column.foreignKey.annotations.contains(module._annotations.FOREIGN_KEY)){
+                var filterPattern = column.foreignKey.annotations.get(module._annotations.FOREIGN_KEY).content.domain_filter_pattern;
+                var uriFilter = module._renderTemplate(filterPattern, data);
+                // NOTE: should we check for (uriFilter.trim() !== '') ?
```
From what I remember we don't need to verify the filter. @hongsudt Do we need to have any sanity checks for the `domain_filter_pattern`?

jrchudy replied:
I don't think `module._renderTemplate` can return anything but the defined string from the filter pattern with the variables replaced with values or `null`. But I wanted to leave that comment there and find out because I wasn't sure.

jrchudy commented:
There are test cases for this in `chaise`. Because it relies on data being defined for the `domain-filter-pattern` to work correctly, I figured it would be more useful to test the functionality from the UI. 

I could write tests for this in `ermrestJS`, but I imagine they would be a near duplicate of the functionality in the tests in `chaise`. The only difference I can imagine is that the test in `ermrestJS` would involve me checking the returned `uri` rather than reading from that set.

As I'm typing this, I realize it wouldn't hurt to have tests for this in `ermrestJS`. I will add these in the morning tomorrow.

jrchudy commented:
I updated the PR with a change that moved filterRef to the `ReferenceColumn` object. I also added a  set of tests to verify that the function behaves properly.

RFSH commented:
Thanks for the changes. It's ready to merge.
